### PR TITLE
KAFKA-5671 Followup: Remove reflections in unit test classes

### PR DESCRIPTION
**Pull request by guozhangwang**
1. Remove rest deprecation warnings in streams:jar.

2. Consolidate all unit test classes' reflections to access internal topology builder from packages other than `o.a.k.streams`. We need to refactor the hierarchies of StreamTask, StreamThread and KafkaStreams to remove these hacky reflections.

3. Minor fixes such as reference path, etc.

4. Minor edits on web docs for the describe function under developer-guide.

### CODE REVIEW

**Comment by guozhangwang** [comment]
@dguy @bbejeck @mjsax  to review.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6453/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6438/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6458/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6443/
Test PASSed (JDK 8 and Scala 2.12).

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java [hunk 1]
```diff
@@ -51,17 +50,17 @@
  * @deprecated Use {@link org.apache.kafka.streams.StreamsBuilder StreamsBuilder} instead
  */
 @Deprecated
-public class KStreamBuilder extends TopologyBuilder {
+public class KStreamBuilder extends org.apache.kafka.streams.processor.TopologyBuilder {
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:53]: why this change?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java [hunk 2]
```diff
@@ -55,13 +54,13 @@ public RecordCollector recordCollector() {
...
     /**
-     * @throws TopologyBuilderException if an attempt is made to access this state store from an unknown node
+     * @throws org.apache.kafka.streams.errors.TopologyBuilderException if an attempt is made to access this state store from an unknown node
      */
     @SuppressWarnings("deprecation")
     @Override
     public StateStore getStateStore(final String name) {
         if (currentNode() == null) {
-            throw new TopologyBuilderException("Accessing from an unknown node");
+            throw new org.apache.kafka.streams.errors.TopologyBuilderException("Accessing from an unknown node");
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java:63]: Again, why?

File: streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java [hunk 3]
```diff
@@ -78,4 +72,24 @@ public static Properties minimalStreamsConfig() {
         return results;
     }
 
+    // TODO: these three static functions are added because some non-TopologyBuilder unit tests need to access the internal topology builder,
+    //       which is usually a bad sign of design patterns between TopologyBuilder and StreamThread. We need to consider getting rid of them later
+    public static InternalTopologyBuilder internalTopologyBuilder(final StreamsBuilder streamsBuilder) throws NoSuchFieldException, IllegalAccessException {
```
Comment (dguy) [inline streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java]: Here and the next method. Can we wrap them with:
```
try {
.../
}catch(final Exception e) {
  throw new RuntimeException(e);
}
```
We can then get rig of the `throws Exception` from all the tests and from the `KStreamTestDriver`

Comment (ijuma) [reply streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java]: Do we really need to use reflection? Can we not add a test class in the same package that exposes a package private field as public? When exploring the codebase is very useful to be able to see callers and reflection kills that.

Comment (dguy) [reply streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java]: @ijuma that is possible. We'd need a couple of test helper classes in the appropriate packages, one for `StreamBuilder` and one for `InternalStreamsBuilder`. We'd also need to make `internalTopologyBuilder` on `StreamsBuilder` package private.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:53]: To avoid importing the `org.apache.kafka.streams.processor.TopologyBuilder` which will introduce a WARNING during compilation; instead we will explicitly refer to the class within the class whenever we have to.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java (hunk 2 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorContextImpl.java:63]: Ditto above.

**Comment by guozhangwang** [comment]
@dguy Comments addressed, could you review again?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6460/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6475/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by bbejeck** [comment]
@guozhangwang @dguy:  sorry I'm a little late with this but I was thinking instead of a static helper class we used a wrapper class instead, something like:
```
 package org.apache.kafka.streams;

 public final class TestingStreamsBuilderWrapper extends StreamsBuilder {

    private final StreamsBuilder delegate;

    public TestingStreamsBuilderWrapper(StreamsBuilder delegate) {
        this.delegate = delegate;
    }

    public InternalTopologyBuilder getInternalTopologyBuilder() {
         return delegate.internalTopologyBuilder;
    }

    public InternalStreamsBuilder getInternalStreamsBuilder() {
        return delegate.internalStreamsBuilder;
    }
```
IMHO using this as a drop in replacement for `StreamsBuilder` might make the current tests easier to follow and subsequent tests easier to write.

**Comment by guozhangwang** [comment]
retest this please.

**Comment by guozhangwang** [comment]
@bbejeck The issue is that `internalTopologyBuilder` is a package-private field of `InternalStreamsBuilder` and `internalStreamsBuilder` is a package-private field of `StreamsBuilder`, which are in `o.a.k.streams.kstream.internals` and `o.a.k.streams` respectively. So if your helper class in only in `o.a.k.streams` it will not be able to access the first field.

**Comment by bbejeck** [comment]
@guozhangwang that's right, we'd need to add an additional helper class, let me know if you think the idea has legs, otherwise I'll just leave it as is.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6481/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by guozhangwang** [comment]
@bbejeck In my latest commit I'm doing similar things but in an existing class than adding additional helper classes. LMK if you find that's good enough.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6466/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6486/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6471/
Test FAILed (JDK 8 and Scala 2.12).

**Review by dguy — APPROVED** [review]
Thanks @guozhangwang, LGTM. 
I wonder if we should move the utility methods to get the `InternalTopologyBuilder` etc onto `XXUtils` classes rather than having them on specific Test classes? I don't have a strong opinion on this, so i'll leave it up to you to decide. Feel free to merge as is

**Comment by dguy** [comment]
@bbejeck that is what i was thinking, but we need to have at least a couple of helper classes in the appropriate packages (which is fine). I'm ok with how it is, though.  TBH i'm not a big fan of either approach, but it is where we are at for the time being.

**Comment by dguy** [comment]
retest this please

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6493/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6478/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by bbejeck** [comment]
@guozhangwang yeah that works for me. 

LGTM

**Comment by bbejeck** [comment]
@dguy agreed on both approaches, but for now we're not relying on reflection which is the important part IMHO.

**Comment by guozhangwang** [comment]
I think it is okay to keep it in the specific test class than creating two more XXTestUtil classes.

**Comment by guozhangwang** [comment]
Merged to trunk.
